presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuilder.java

```diff
@@ -270,7 +271,10 @@ public void addInput(Page page)
                 size += index.getEstimatedSize().toBytes();
             }
 
-            operatorContext.setMemoryReservation(size);
+            // PagesIndex.addPage doesn't track page memory but blocks are shared and kept
+            pagesMemorySize += page.getRetainedSizeInBytes();
```

Comment:
Nevertheless calculating memory used by PageIndex here seems like not appropriate place - it is easy to miss it during refactor and every user of PageIndex will have to take care about it. 

I agree that it seems tricky. But maybe there should be some more intelligent mechanism for that? For example user would "create memory reservation" for whole page. Memory manager would return unique id for such memory reservation and it would hold reference counter, which user would be able to increase/decrease. Memory usage for such Page, that is being shared between many operators, would also be shared. Ownership would be transferred without glitches and race conditions between following operators and it would be released once last user/operator of the given memory reservation destroys it/doesn't need it.

What do you think? 

I'm not sure if we can submit this commit as it is with this race condition - it can lead to random query failing, that's are barely at the memory limit.